File: backend/app/schemas/library_audit.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class SimilarNameMember(BaseModel):
    id: int
    name: str
    book_count: int
    file_count: int


class SimilarNameCluster(BaseModel):
    members: list[SimilarNameMember]


class SimilarCoverPair(BaseModel):
    """Two organised files that resolved to different books but whose cover
    thumbnails are within a small perceptual-hash distance — likely the same
    book re-uploaded with rewritten metadata (which sha256 dedup misses)."""

    book_a_id: int
    book_a_title: str
    file_a_name: str
    book_b_id: int
    book_b_title: str
    file_b_name: str
    distance: int


class LibraryAuditResult(BaseModel):
    similar_series: list[SimilarNameCluster]
    similar_authors: list[SimilarNameCluster]
    similar_covers: list[SimilarCoverPair] = []


class DismissClusterRequest(BaseModel):
    kind: Literal["series", "author"]
    member_ids: list[int]


class DismissedClusterInfo(BaseModel):
    id: int
    kind: Literal["series", "author"]
    member_ids: list[int]
    created_at: datetime


class TitleMergeRepairResult(BaseModel):
    books_split: int
    files_moved: int

File: frontend/src/types/libraryAudit.ts
export interface SimilarNameMember {
  id: number
  name: string
  book_count: number
  file_count: number
}

export interface SimilarNameCluster {
  members: SimilarNameMember[]
}

export interface SimilarCoverPair {
  book_a_id: number
  book_a_title: string
  file_a_name: string
  book_b_id: number
  book_b_title: string
  file_b_name: string
  distance: number
}

export interface LibraryAuditResult {
  similar_series: SimilarNameCluster[]
  similar_authors: SimilarNameCluster[]
  similar_covers: SimilarCoverPair[]
}

export type AuditClusterKind = 'series' | 'author'

export interface DismissedClusterInfo {
  id: number
  kind: AuditClusterKind
  member_ids: number[]
  created_at: string
}

export interface TitleMergeRepairResult {
  books_split: number
  files_moved: number
}
